File: src/main/java/com/refinedmods/refinedstorage/tile/data/TileDataParameter.java
package com.refinedmods.refinedstorage.tile.data;

import net.minecraft.network.syncher.EntityDataSerializer;
import net.minecraft.world.level.block.entity.BlockEntity;

import javax.annotation.Nullable;
import java.util.function.BiConsumer;
import java.util.function.Function;

public class TileDataParameter<T, E extends BlockEntity> {
    private final EntityDataSerializer<T> serializer;
    private final Function<E, T> valueProducer;
    @Nullable
    private final BiConsumer<E, T> valueConsumer;
    @Nullable
    private final TileDataParameterClientListener<T> listener;
    private int id;
    private T value;

    public TileDataParameter(EntityDataSerializer<T> serializer, T defaultValue, Function<E, T> producer) {
        this(serializer, defaultValue, producer, null);
    }

    public TileDataParameter(EntityDataSerializer<T> serializer, T defaultValue, Function<E, T> producer, @Nullable BiConsumer<E, T> consumer) {
        this(serializer, defaultValue, producer, consumer, null);
    }

    public TileDataParameter(EntityDataSerializer<T> serializer, T defaultValue, Function<E, T> producer, @Nullable BiConsumer<E, T> consumer, @Nullable TileDataParameterClientListener<T> listener) {
        this.value = defaultValue;
        this.serializer = serializer;
        this.valueProducer = producer;
        this.valueConsumer = consumer;
        this.listener = listener;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public EntityDataSerializer<T> getSerializer() {
        return serializer;
    }

    public Function<E, T> getValueProducer() {
        return valueProducer;
    }

    @Nullable
    public BiConsumer<E, T> getValueConsumer() {
        return valueConsumer;
    }

    public void setValue(boolean initial, T value) {
        this.value = value;

        if (listener != null) {
            listener.onChanged(initial, value);
        }
    }

    public T getValue() {
        return value;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/tile/data/TileDataParameter.kt
package com.refinedmods.refinedstorage.tile.data

import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.data.TrackedDataHandler
import java.util.function.BiConsumer

class TileDataParameter<T: Any?, E : BlockEntity?>(
        var value: T?,
        val serializer: TrackedDataHandler<T?>,
        val valueProducer: java.util.function.Function<E?, T?>,
        val valueConsumer: BiConsumer<E?, T?>? = null,
        val listener: TileDataParameterClientListener<T?>? = null
) {
    var id = 0

    fun setValue(initial: Boolean, value: T?) {

        this.value = value
        listener?.onChanged(initial, value)
    }

}